### ISSUE TITLE
A mistake in positional encoding

### ISSUE THREAD

**Issue by iclementine** [janvainer/speedyspeech#23]
The positional encoding used in the code is

https://github.com/janvainer/speedyspeech/blob/3a30645249c9f60b9655626afbb263f3e43a4758/code/functional.py#L43

I think it is intended to use the positional encoding in __Attention is all you need__ and scaled by a positioning factor like in deep voice 3. If so, then there is a bug here.

<img width="285" alt="图片" src="https://user-images.githubusercontent.com/16222986/108679465-8924d980-7527-11eb-8c6a-853cdca73a84.png">

`a ** b ** c` is actually `a ** (b**c)` instead of `(a ** b)** c`

![图片](https://user-images.githubusercontent.com/16222986/108679394-6bf00b00-7527-11eb-90f6-6aaec7df5221.png)

**Comment by janvainer** [comment]
Hi @iclementine. I think you are right and this is really a bug. I would love to merge your PR, but right now, this would break compatibility of the code with the pretrained checkpoints. The SpeedySpeech checkpoint would be ok though, only the duration predictor would break. The positional encodings right now work fine (the bug stretches both encodings by the same amount so the alignment is still correct). I created a development branch, where I will be fixing more bugs and add more functionality. Would you mind changing your PR to address the develop branch? I will release the bugfix along with more changes also with re-trained checkpoints.

**Comment by iclementine** [comment]
ok, thank you!

**Comment by janvainer** [comment]
Great, approved and merged!

**Comment by janvainer** [comment]
@iclementine are there any ideas/suggestions that would help you get more from this repo? I will be adding support for more recent vocoders, better and more generic data pipelines and fix a traning bug that some users encountered, but I would love to hear some additional thoughts regarding what would be useful for people that use/would like to use this project.

**Comment by iclementine** [comment]
I am impressed by the model's speed and robustness at synthesizing long utterances. I used it to synthesize several paragraph in The Grate Gatsby, among which some utterance is about 20 ~ 30 seconds, which is longer than all the utterances in LJSpeech. The synthesized audio is alright in alignment (thought abit unnatural or influent).

Also I have a few questions:
1. For how many steps is the released model trained. Are they trained for the preset epochs in the training script?

I was working on porting this model for multispeaker TTS and GST-enhanced Expressive TTS. And I have other considerations about the model and data processing.

1. I think remocing stop token prediction is a good idea, since it's too sensitive to silence trimming and splitting of the utteramnces. Explicit duration modeling get rid of the problem of imbalanced [stop/non-stop] classification once for all.
2. Imposing a guided attention loss seems to assume that the audio is trimmed well. If there is long silence in the begining or end of the utterance, the alignment may be a polyline and deviate from the diagonal line.
3. Since the duration of the utterance is the summation of the durations of each phoneme, how should the pauses handled. Some long pauses may have their counterpart in text as punctuations, but if the training dataset have some unmarked pauses, how can we deal with that.

**Comment by janvainer** [comment]
Hi @iclementine. The released model is trained with the default config.
Thanks you for the considerations. I generally agree with the first point. Regarding the second point, what you are suggesting is probably a good idea. The last point could be solved by using <bos> and <eos> tokens for beginning and end silence segments, but that can be solved by trimming too. Silence in the middle of the utterance should map to a <space> character. We could also consider some blank tokens like in CTC loss, but that would make thinks pretty complicated.
